File: 프로그래머스/lv2/42577. 전화번호 목록/전화번호 목록.java
import java.util.*;
class Solution {
    public boolean solution(String[] phone_book) {
        Map<String,String> map = new HashMap<>();
        for(String phone: phone_book){
            map.put(phone,phone);
        }
        for(String phone: phone_book){
            for(int end=1; end<=phone.length();end++){
                if(
                    map.containsKey(phone.substring(0,end)) &&
                    !phone.equals(phone.substring(0,end))
                )
                    return false;
            }   
        }
        return true;
    }
}

File: 프로그래머스/lv2/42577. 전화번호 목록/전화번호 목록.py
from collections import defaultdict
def solution(phone_book):
    if len(phone_book) == 1:
        return True
    lengths = set([len(phone) for phone in phone_book])
    di = defaultdict(int)
    for phone in phone_book:
        for length in lengths:
            if len(phone) >= length:
                di[phone[:length]]+=1
    for phone in phone_book:
        if di[phone] >= 2:
            return False
    return True